app/models/action.rb

```diff
@@ -0,0 +1,15 @@
+class Action < ApplicationRecord
+  enum status: [:pending, :completed, :failed]
+
+  belongs_to :workflow
+  belongs_to :subject
+
+  def perform
+    effect.perform(workflow_id, subject_id)
```

Comment:
As a future enhancement, we could consider passing the `Action` itself into the `Effect` so that the `Effect` will be able to mark the `Action` as failed under circumstances where it doesn't have to throw an exception.